### PR TITLE
tests: stabilize test_sharding_split_compaction

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

This test incorrectly assumed that a post-split compaction would only drop content.  This was easily destabilized by any changes to image generation rules.

In this, and other tests that explicitly create image layers (e.g. latest test_pg_regress & test_isolation), there is also a bug in compaction with force-image-layers, where it will fail trying to create a duplicate image layer.

## Summary of changes

- Before split, do a full image layer generation pass, to guarantee that post-split compaction should only drop data, never create it.
- Fix the force_image_layer_creation mode of compaction that we use from tests like this: previously it would try and generate image layers even if one already existed with the same layer key, which caused compaction to fail.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: test_runner/regress/test_sharding.py [hunk 1]
```diff
@@ -225,6 +225,12 @@ def test_sharding_split_compaction(neon_env_builder: NeonEnvBuilder, failpoint:
     workload.validate()
     workload.stop()
 
+    # Do a full image layer generation before splitting, so that when we compact after splitting
+    # we should only see sizes decrease (from post-split drops/rewrites), not increase (from image layer generation)
+    env.get_tenant_pageserver(tenant_id).http_client().timeline_compact(
+        tenant_id, timeline_id, force_image_layer_creation=True
```
Comment (koivunej) [inline test_runner/regress/test_sharding.py:231]: This needs to await uploads, I added an arg wait uploads.

Comment (koivunej) [reply test_runner/regress/test_sharding.py:231]: We discussed this in the call just now, you had the idea that perhaps as part of shard splitting the parent should always be awaited to complete remote ops, or flushed, I forgot which already.

Comment (jcsp) [reply test_runner/regress/test_sharding.py]: I didn't know about `wait_until_uploaded` -- add that here.

**Review by skyzh — APPROVED** [review]

File: pageserver/src/tenant/timeline.rs [hunk 2]
```diff
@@ -4572,6 +4573,22 @@ impl Timeline {
...
                 }
+            } else if let ImageLayerCreationMode::Force = mode {
+                // When forced to create image layers, we might try and create them where they already
+                // exist.  This mode is only used in tests/debug.
+                let layers = self.layers.read().await;
+                if layers.contains_key(&PersistentLayerKey {
+                    key_range: img_range.clone(),
+                    lsn_range: PersistentLayerDesc::image_layer_lsn_range(lsn),
+                    is_delta: false,
+                }) {
```
Comment (koivunej) [inline pageserver/src/tenant/timeline.rs:4584]: We are doing this query against the hashmap; this only works assuming the partitioning doesn't change. Should the query be done against layermap and image coverage?

Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:4584]: This is `#[cfg(feature = testing)]` afaik, so maybe something to consider. I do wonder if this means a bug with the overall force handling.

**Pull request merged**